core/src/main/java/com/linecorp/armeria/internal/common/util/TemporaryThreadLocals.java

```diff
@@ -15,34 +15,65 @@
  */
 package com.linecorp.armeria.internal.common.util;
 
+import static com.google.common.base.Preconditions.checkState;
+
 import com.google.common.annotations.VisibleForTesting;
 
 import io.netty.util.internal.EmptyArrays;
 
 /**
  * Provides various thread-local variables used by Armeria internally, mostly to avoid allocating
  * short-living objects, such as {@code byte[]} and {@link StringBuilder}. Keep in mind that the variables
- * provided by this class must be used with extreme care, because otherwise it will result in unpredictable
- * behavior.
+ * provided by this class must be used and released with extreme care, because otherwise it will result in
+ * unpredictable behavior.
  *
  * <p>Most common mistake is to call or recurse info a method that uses the same thread-local variable.
  * For example, the following code will produce a garbled string:
  * <pre>{@code
  * > class A {
  * >     @Override
  * >     public String toString() {
- * >         return TemporaryThreadLocals.get().append('"').append(new B()).append('"').toString();
+ * >         StringBuilder stringBuilder = TemporaryThreadLocals.get().stringBuilder();
+ * >         TemporaryThreadLocals.get().releaseStringBuilder();
+ * >         return stringBuilder.append('"').append(new B()).append('"').toString();
  * >     }
  * > }
  * > class B {
  * >     @Override
  * >     public String toString() {
- * >         return TemporaryThreadLocals.get().append("foo").toString();
+ * >         StringBuilder stringBuilder = TemporaryThreadLocals.get().stringBuilder();
+ * >         TemporaryThreadLocals.get().releaseStringBuilder();
+ * >         return stringBuilder.append("foo").toString();
  * >     }
  * > }
  * > // The following assertion fails, because A.toString() returns "foofoo\"".
  * > assert "\"foo\"".equals(new A().toString());
- * }</pre></p>
+ * }</pre>
+ *
+ * <p>A release method is helpful to not only prevent from being corrupted but also recognize the situation
+ * about nested use. Specifically, in case of {@link StringBuilder}, {@code toString()} is expected prior to
+ * {@code releaseStringBuilder()}:
+ * <pre>{@code
+ * > class A {
+ * >     @Override
+ * >     public String toString() {
+ * >         StringBuilder stringBuilder = TemporaryThreadLocals.get().stringBuilder();
+ * >         String toString = stringBuilder.append('"').append(new B()).append('"').toString();
+ * >         threadLocals.get().releaseStringBuilder();
```

Comment:
Had a chat with @trustin and @ikhoon about the API design.
We usually use try-with-resources for releasing the resources instead of manually calling the method such as `releaseStringBuilder()`.

So, How about using the same approach here?
```java
try (ThreadLocalVariables tlvs = TemporaryThreadLocals.acquire()) {
   tlvs.stringBuffer();
   ...
}
// ThreadLocalVariables implements SafeCloseable.
```

Reply:
Sounds good to me. Then I have two questions.
1. According to your example, `ThreadLocalVariables` is returned. It seems to have all shared resources including int array, byte array and so on. What do you think about declaring resource classes each such as `ThreadLocalIntArray` and `ThreadLocalStringBuilder`? because there is no need to have a same lock across resources. For example, int array and string builder can be used independently.
2. In light of try-with-resources statement, if `acquire()` throws an exception, nested use will be no-op. That is, there is no corruption. However, as the exception is ignored silently and suppressed, recognizing the no-op by nested use may be hard. 🤔 If caller is in charge of fallback in this API design, is there no need to consider as follows?
```java
try (ThreadLocalStringBuilder stringBuilder = TemporaryThreadLocals.acquireStringBuilder()) {
    // something to do
    return stringBuilder.toString();
}
// fallback
return "";
```

Reply:
> For example, int array and string builder can be used independently.

Can't we just use them in the same block?

>  if acquire() throws an exception, nested use will be no-op.

Isn't it the same when `TemporaryThreadLocals.get().stringBuilder();` throws an exception?

Reply:
> Can't we just use them in the same block?

We can. My point was giving more freedom to users according to resources. I have no concern or preference. I just wonder your preference.

> throws an exception?

Same. By the way, I would like to discuss about one of my suggestions helping to recognize the nested use. Current code or `acquire()` with try-with-resources ignores the unexpected behavior silently. What do you think about adding the logging when `TemporaryThreadLocalVariables` are used in this time?
```java
try (ThreadLocalVariables tlvs = TemporaryThreadLocals.acquire()) {
    tlvs.stringBuffer();
    ...
} catch (RuntimeException e /* or new exception for nested use? */) {
    logger.log(..., e);
}
```

Reply:
> We can. My point was giving more freedom to users according to resources. I have no concern or preference. I just wonder your preference.

Well, we can provide separate methods for acquiring `StringBuilder`, `byte[]`, `char[]` and `int[]`.
There's no problem with it, but we have to have four separate classes for that. If we need to add some more stuff such as `long[]`, then we need to add an additional class.
But if we just use `ThreadLocalVariables`, we can just add a method.
I prefer adding a method instead of adding a class. That's my preference. 😄

> acquire() with try-with-resources ignores the unexpected behavior silently

I think I didn't get this point. 😅 I think this does not ignore the exception unless we swallow the exception using `catch`. Could you show me an example of the nested case you talked about, please? 😄

Reply:
> but we have to have four separate classes for that. If we need to add some more stuff such as long[], then we need to add an additional class.

Exactly. So I also concerned "more freedom vs. maintenance cost from more codes". I see. Thanks for sharing the preference. I will update in that way.

> unless we swallow the exception using catch

Understood what you meant. I thought `catch` will be used basically, but it is not. As you mentioned, most statement uses `try` solely like https://github.com/line/armeria/blob/master/brave/src/main/java/com/linecorp/armeria/client/brave/BraveClient.java#L115.

Reply:
> Thanks for sharing the preference. I will update in that way.

Thanks for understanding. 😄 

> I thought catch will be used basically, but it is not. As you mentioned, most statement uses try solely like

Yeah, even though `catch` is used it's no problem unless it just swallows the exception. (I didn't see any that sort of code in Armeria. 😄 )

Reply:
@trustin According to https://github.com/line/armeria/pull/3505#issuecomment-849417310, it seems similar with my first proposal and different with the discussion in this thread. I thought @minwoox 's opinion is same with @trustin and @ikhoon from https://github.com/line/armeria/pull/3505#discussion_r635822079. What do we prefer? `AutoClosable` for try-with-resources vs. try-finally.

Reply:
Ah, `TemporaryThreadLocal` implementing `AutoCloseable` sounds good to me. +1 for `AutoCloseable`.

Reply:
Ah, yeah, we can use `TemporaryThreadLocal` and it shouldn't be `SafeCloseable` but `AutoCloseable`. 😅